### PR TITLE
Simplified BatchSpanProcessor

### LINKED ISSUE #729: Simplify the BatchSpansProcessor
I believe this class can be re-implemented to use a simple Queue and a background thread to process it.

### PR DESCRIPTION
**Pull request by neeraj97**
Fixes #729 
Used ScheduledExecutorService to schedule background worker to export spans with fixed delays and also forced to execute the worker if the queue size is greater than the halfMaxQueueSize and ScheduledExecutorService  is idle. BlockingQueue is used to avoid manual locks.

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 1]
```diff
@@ -88,8 +91,10 @@
   private static final String EXPORTER_THREAD_NAME =
       BatchSpanProcessor.class.getSimpleName() + "_ExporterThread";
   private final Worker worker;
-  private final Thread workerThread;
   private final boolean sampled;
+  private final ScheduledExecutorService scheduler;
+  private final ScheduledFuture<?> scheduledFutureWithFixedDelay;
+  private Future<?> imperativeSubmission;
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:97]: I think this reference needs to either be volatile, or implemented with an AtomicReference, since it's assignment needs to be visible across threads.

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:97]: Also, maybe renamed this to "forcedExportResult" or something along those lines? Not sure exactly what the best name is, but th
is name tripped me up for a bit.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 2]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
-      this.scheduleDelayMillis = scheduleDelayMillis;
       this.maxQueueSize = maxQueueSize;
       this.halfMaxQueueSize = maxQueueSize >> 1;
       this.maxExportBatchSize = maxExportBatchSize;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.spansList = new ArrayBlockingQueue<>(maxQueueSize);
       this.exporterTimeoutMillis = exporterTimeoutMillis;
     }
 
+    private static List<SpanData> createSpanDataForExport(
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:217]: Can you move this back down below where it's used?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 3]
```diff
@@ -98,16 +103,22 @@ private BatchSpanProcessor(
...
+  }
+
+  /**
+   * Returns a new Builder for {@link BatchSpanProcessor}.
+   *
+   * @param spanExporter the {@code SpanExporter} to where the Spans are pushed.
+   * @return a new {@link BatchSpanProcessor}.
+   * @throws NullPointerException if the {@code spanExporter} is {@code null}.
+   */
+  public static Builder newBuilder(SpanExporter spanExporter) {
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:120]: Why move this method as a part of this PR?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 4]
```diff
@@ -124,6 +135,17 @@ public void onEnd(ReadableSpan span) {
       return;
     }
     worker.addSpan(span);
+
+    if (worker.spansList.size() >= worker.halfMaxQueueSize) {
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:139]: Can this logic live down in the worker itself, rather than spreading it across two places?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 5]
```diff
@@ -98,16 +103,22 @@ private BatchSpanProcessor(
...
-            spanExporter,
-            scheduleDelayMillis,
-            maxQueueSize,
-            maxExportBatchSize,
-            exporterTimeoutMillis);
-    this.workerThread = new DaemonThreadFactory(WORKER_THREAD_NAME).newThread(worker);
-    this.workerThread.start();
+    this.worker = new Worker(spanExporter, maxQueueSize, maxExportBatchSize, exporterTimeoutMillis);
     this.sampled = sampled;
+    scheduler = Executors.newScheduledThreadPool(1, new DaemonThreadFactory(WORKER_THREAD_NAME));
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:108]: `newSingleThreadScheduledExecutor`

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 1 above)
Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:97]: From what I can tell, it's pretty easy to schedule something immediately many times because of the thread safety which will be hard to solve. I think instead of a future, we can just use an `AtomicBoolean`

```java
onEnd() {
  if (forcingExport.compareAndSet(false, true)) {
    scheduler.submit(worker);
  }
}

...

Worker.run() {
...

// Last line
forcingExport.compareAndSet(true, false);
```

Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:97]: That being said, this forced export logic seems suspicious. If the queue isn't keeping up, then it seems the worker will run consecutively forever ignoring `exportTimeoutMillis` completely because `worker.spansList.size() >= worker.halfMaxQueueSize` will always be true any time `forcingExport` is false in the pathological case. I think users would rather have spans dropped than this heavyweight thread running all the time?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 6]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
-          return;
-        }
-        // TODO: Record a gauge for referenced spans.
-        spansList.add(span);
-        // Notify the worker thread that at half of the queue is available. It will take
-        // time anyway for the thread to wake up.
-        if (spansList.size() >= halfMaxQueueSize) {
-          monitor.notifyAll();
-        }
+      if (spansList.size() == maxQueueSize) {
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:229]: This doesn't look threadsafe

```
if (!spansList.offer(span)) {
  droppedSpans.add(1);
}
```

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 7]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
-    private final long scheduleDelayMillis;
     private final int maxQueueSize;
     private final int maxExportBatchSize;
     private final int halfMaxQueueSize;
-    private final Object monitor = new Object();
     private final int exporterTimeoutMillis;
-
-    @GuardedBy("monitor")
-    private final List<ReadableSpan> spansList;
+    private final BlockingQueue<ReadableSpan> spansList;
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:202]: It's not a list it's a queue, rename to just `spans` to prevent confusion

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 8]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
+      this.spansList = new ArrayBlockingQueue<>(maxQueueSize);
       this.exporterTimeoutMillis = exporterTimeoutMillis;
     }
 
+    private static List<SpanData> createSpanDataForExport(
+        List<ReadableSpan> spanList, int startIndex, int numberToTake) {
+      List<SpanData> spanDataBuffer = new ArrayList<>(numberToTake);
+      for (int i = startIndex; i < numberToTake; i++) {
+        spanDataBuffer.add(spanList.get(i).toSpanData());
+        // Remove the reference to the ReadableSpan to allow GC to free the memory.
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:222]: Hmm, since the list will be GC'd at the same time as the items, I don't think this actually helps

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 9]
```diff
@@ -247,11 +251,10 @@ private void shutdown() {
     }
 
     private void forceFlush() {
-      ArrayList<ReadableSpan> spansCopy;
-      synchronized (monitor) {
-        spansCopy = new ArrayList<>(spansList);
-        spansList.clear();
-      }
+      // Draining all the batched spans in a separate list
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:254]: Not sure if we really need this comment, it's obvious for blocking queues

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 10]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
       this.maxQueueSize = maxQueueSize;
       this.halfMaxQueueSize = maxQueueSize >> 1;
       this.maxExportBatchSize = maxExportBatchSize;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.spansList = new ArrayBlockingQueue<>(maxQueueSize);
       this.exporterTimeoutMillis = exporterTimeoutMillis;
     }
 
+    private static List<SpanData> createSpanDataForExport(
+        List<ReadableSpan> spanList, int startIndex, int numberToTake) {
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:218]: Fix `numberToTake` to `endIndex`

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 11]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
+      List<SpanData> spanDataBuffer = new ArrayList<>(numberToTake);
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:219]: Due to th edeceptive name `numberToTake`, this looks like a bug, the list gets bigger and bigger and bigger :)

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 12]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
-              }
-            } while (spansList.isEmpty());
-          }
-          spansCopy = new ArrayList<>(spansList);
-          spansList.clear();
-        }
-        // Execute the batch export outside the synchronized to not block all producers.
-        exportBatches(spansCopy);
-      }
+      // Draining all the batched spans in a separate list
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:239]: Ditto

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 13]
```diff
@@ -163,81 +192,56 @@ public void forceFlush() {
...
+      // avoid blocking the producer thread.
+      ArrayList<ReadableSpan> spansCopy = new ArrayList<>();
+      spansList.drainTo(spansCopy);
+      // Execute the batch export outside the synchronized to not block all producers.
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:243]: No more synchronized, remove

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 9 above)
Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:254]: Actually, looks like we can remove this method now it is identical to run

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 14]
```diff
@@ -303,32 +295,19 @@ public void run() {
...
-    private static final String KEY_SAMPLED = "otel.bsp.export.sampled";
-
     @VisibleForTesting static final long DEFAULT_SCHEDULE_DELAY_MILLIS = 5000;
     @VisibleForTesting static final int DEFAULT_MAX_QUEUE_SIZE = 2048;
     @VisibleForTesting static final int DEFAULT_MAX_EXPORT_BATCH_SIZE = 512;
     @VisibleForTesting static final int DEFAULT_EXPORT_TIMEOUT_MILLIS = 30_000;
     @VisibleForTesting static final boolean DEFAULT_EXPORT_ONLY_SAMPLED = true;
-
+    private static final String KEY_SCHEDULE_DELAY_MILLIS = "otel.bsp.schedule.delay";
+    private static final String KEY_MAX_QUEUE_SIZE = "otel.bsp.max.queue";
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:307]: Why did you change this ordering? We usually have private before non-private

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 1 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:97]: I think this is a good question to ask. I believe it applies to this class on the main branch as well. Maybe this should be addressed first? Can you create an issue for it?

[also, yes, just using an AtomicBoolean seems like a much simpler solution if we *do* want to force exports when the queue is full, before the interval is complete]

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 10 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:218]: I think this is the number to take if you look how it's used in the exportBatches method.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 8 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:222]: Agreed. I think this is a holdover from the previous implementation.

**Comment by jkwatson** [comment]
There are at least 2 tests that are broken with this implementation. I would hope that the tests would all still pass with a change just in the implementation, rather than changing the behavior.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java [hunk 15]
```diff
@@ -98,16 +103,22 @@ private BatchSpanProcessor(
...
-            maxQueueSize,
-            maxExportBatchSize,
-            exporterTimeoutMillis);
-    this.workerThread = new DaemonThreadFactory(WORKER_THREAD_NAME).newThread(worker);
-    this.workerThread.start();
+    this.worker = new Worker(spanExporter, maxQueueSize, maxExportBatchSize, exporterTimeoutMillis);
     this.sampled = sampled;
+    scheduler = Executors.newScheduledThreadPool(1, new DaemonThreadFactory(WORKER_THREAD_NAME));
+    scheduledFutureWithFixedDelay =
+        scheduler.scheduleWithFixedDelay(worker, 0, scheduleDelayMillis, TimeUnit.MILLISECONDS);
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:110]: I don't think you want to schedule an export immediately on startup, do you? I'd recommend delaying by the specified delay before the first export.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java (hunk 10 above)
Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:218]: Really? We iterate from startIndex to this number in the loop

`for (int i = startIndex; i < numberToTake; i++) {`

If it was actually the number to take, it'd be 

`for (int i = startIndex; i < i + numberToTake; i++) {`

wouldn't it?

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:218]: yikes! This is a really bad bug. Thanks for the sharp eye on that! Good thing no one has been trying to send too many spans with this thing yet!

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:218]: ok, it wasn't really all that bad a bug, but @anuraaga is correct. The parameter is misnamed, both here and in the caller. This has been fixed on the main branch with #1400 , so if you could fix it here as well, that would be good.

**Comment by neeraj97** [comment]
@jkwatson @anuraaga Thank you for the detailed review. I got some amazing insights into problem. I will work on them.
